server/vitals.js

```diff
@@ -183,28 +183,27 @@ async function handleHeartbeat(req, res) {
         // Must send 200 so as not to be throttled by Particle (ref: https://docs.particle.io/reference/device-cloud/webhooks/#limits)
         res.status(200).json(errorMessage)
       } else {
-        const currentDbTime = await db.getCurrentTime()
-
         const message = JSON.parse(req.body.data)
         const doorMissedMessagesCount = message.doorMissedMsg
         const doorLowBatteryFlag = message.doorLowBatt
-        const doorTimeSinceLastHeartbeat = new Date(currentDbTime)
-        doorTimeSinceLastHeartbeat.setMilliseconds(currentDbTime.getMilliseconds() - message.doorLastHeartbeat)
         const resetReason = message.resetReason
         const stateTransitionsArray = message.states.map(convertStateArrayToObject)
 
         if (doorLowBatteryFlag) {
           await sendLowBatteryAlert(location.locationid)
         }
 
-        await db.logSensorsVital(
-          location.locationid,
-          doorMissedMessagesCount,
-          doorLowBatteryFlag,
-          doorTimeSinceLastHeartbeat,
-          resetReason,
-          stateTransitionsArray,
-        )
+        let doorLastSeenAt
+        if (message.doorLastMessage.toString() === '-1') {
+          // TODO transaction?
```

Comment:
Put it in a transaction?

Reply:
Decided not to in order to not introduce new places for deadlocks. I think the chance to this function being called twice for the same location in a short enough period of time to cause a race condition is very low.